Orange/classification/rules.py

```diff
@@ -0,0 +1,947 @@
+            x = np.array([x[rule.target_class],
+                          np.sum(x) - x[rule.target_class]])
+        return -entropy(x)
+
+
+class LaplaceAccuracyEvaluator(Evaluator):
+    def evaluate_rule(self, rule):
+        if rule.target_class is not None:
+            target = rule.curr_class_dist[rule.target_class]
+            combined = np.sum(rule.curr_class_dist)
+            k = 2
+        else:
+            target = np.argmax(rule.curr_class_dist)
+            combined = np.sum(rule.curr_class_dist)
+            k = len(rule.curr_class_dist)
+        return (target + 1) / (combined + k)
+
+
+class LengthEvaluator(Evaluator):
+    def evaluate_rule(self, rule):
+        return -rule_length(rule)
+
+
+class Validator:
+    def validate_rule(self, rule):
+        """
+        Characterise a heuristic to avoid the over-fitting of noisy data
+        and to reduce computation time.
+
+        Parameters
+        ----------
+        rule : Rule
+            Validate this rule.
+
+        Returns
+        -------
+        res : bool
+            Validation function result.
+        """
+        raise NotImplementedError("descendants of Validator must "
+                                  "overload method 'validate_rule'")
+
+
+class CustomGeneralValidator(Validator):
+    """
+    Discard rules that
+
+    - cover less than the minimum required number of examples,
+    - offer no additional advantage compared to their parent rule,
+    - are too complex.
+    """
+    def __init__(self, max_rule_length=5, minimum_covered_examples=1):
+        self.max_rule_length = max_rule_length
+        self.minimum_covered_examples = minimum_covered_examples
+
+    def validate_rule(self, rule):
+        rule_target_covered = (rule.curr_class_dist[rule.target_class]
+                               if rule.target_class is not None
+                               else np.sum(rule.curr_class_dist))
+
+        return (rule_target_covered >= self.minimum_covered_examples and
+                rule_length(rule) <= self.max_rule_length and
+                (not np.array_equal(rule.curr_class_dist,
+                                    rule.parent_rule.curr_class_dist)
+                 if rule.parent_rule is not None else True))
+
+
+class LRSValidator(Validator):
+    """
+    To test significance, calculate the likelihood ratio statistic. It
+    provides an information-theoretic measure of the (non-commutative)
+    distance between two distributions. Under suitable assumptions, it
+    can be shown that the statistic is distributed approximately as
+    Chi^2 probability distribution with n-1 degrees of freedom. As the
+    score lowers, the apparent regularity is more likely observed due to
+    chance.
+    """
+    def __init__(self, alpha=1.0):
+        self.alpha = alpha
+
+    def validate_rule(self, rule):
+        if self.alpha >= 1.0 or rule.parent_rule is None:
+            return True
+
+        x = rule.curr_class_dist.astype(dtype=np.float)
```

Comment:
You can use Python's types (int, float) and Numpy does it correctly. I find it more readable without superfluous `np.` everywhere.